codec/src/main/java/io/netty/handler/codec/BaseLengthBasedFrameDecoder.java

```diff
@@ -0,0 +1,103 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.serialization.ObjectDecoder;
+
+import java.util.List;
+
+/**
+ * Encapsulates common logic and state for message decoders that attempt to detect frames
+ * of a certain length
+ */
+public abstract class BaseLengthBasedFrameDecoder extends ByteToMessageDecoder {
+
+    protected final int maxFrameLength;
+    protected final boolean failFast;
+    protected boolean discardingTooLongFrame;
+    protected long tooLongFrameLength;
+    protected long bytesToDiscard;
+
+    protected BaseLengthBasedFrameDecoder(int maxFrameLength, boolean failFast) {
+        this.maxFrameLength = maxFrameLength;
+        this.failFast = failFast;
+    }
+
+    /**
+     * Fail the decoding due to a frame that is too long, as per {@link #maxFrameLength}
+     *
+     * @param firstDetectionOfTooLongFrame true if this is the first time the too-long frame was detected
+     * @return true if all bytes from the too-long frame were discarded
+     */
+    protected boolean failIfNecessary(boolean firstDetectionOfTooLongFrame) {
+        if (bytesToDiscard == 0) {
+            // Reset to the initial state and tell the handlers that
+            // the frame was too large.
+            long tooLongFrameLength = this.tooLongFrameLength;
+            this.tooLongFrameLength = 0;
+            discardingTooLongFrame = false;
+            if (!failFast || firstDetectionOfTooLongFrame) {
+                fail(tooLongFrameLength);
+            }
+            return true;
+        } else {
+            // Keep discarding and notify handlers if necessary.
+            if (failFast && firstDetectionOfTooLongFrame) {
+                fail(tooLongFrameLength);
+            }
+            return false;
+        }
+    }
+
+    protected void resetStateOnTooLongFrame() {
+        // do nothing by default
+    }
+
+    /**
+     * Extract the sub-region of the specified buffer.
+     * <p>
+     * If you are sure that the frame and its content are not accessed after
+     * the current {@link #decode(ChannelHandlerContext, ByteBuf)}
+     * call returns, you can even avoid memory copy by returning the sliced
+     * sub-region (i.e. <tt>return buffer.slice(index, length)</tt>).
+     * It's often useful when you convert the extracted frame into an object.
+     * Refer to the source code of {@link ObjectDecoder} to see how this method
+     * is overridden to avoid memory copy.
+     */
+
+    protected ByteBuf extractFrame(ChannelHandlerContext ctx, ByteBuf buffer, int index, int length) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "ctx". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)